Add core dependencies installation script

ovysotska opened the pull request:
This PR adds a script to pull and install all core dependencies needed for the LaMAR.
Warning, the C++ dependencies will be installed system wide.

This PR also:
* downgrades the version of pycolmap for compatibility issues with COLMAP 3.8.
* downgrades the numpy version in scantools for compatibility with the numpy version of LaMAR.
* adds dependency for scikit-learn because HLOC 1.4 requires it.
* update Dockerfile.

ovysotska commented:
@microsoft-github-policy-service agree

mihaidusmanu commented:
Thanks a lot for the contribution @ovysotska !

The docker image is not fully functional either as some calls to pycolmap in this pipeline are not compatible with 0.6.0.
That package should be downgraded in any case (or someone should make a full run over the pipeline to fix all the compatibility issues with 0.6.0).

I think it's better if users can also install from scratch without needing docker so having a dedicated script for local setup would be a good thing.

mihaidusmanu commented on README.md:
```diff
@@ -68,9 +68,17 @@ T_w_i = sessions.trajectories[keys[0]]  # first pose, from sensor/rig to world
-- Python >= 3.9
-- [hloc v1.4](https://github.com/cvg/Hierarchical-Localization) and its dependencies, including [COLMAP 3.8](https://colmap.github.io/install.html) built from source.
-- [pyceres](https://github.com/cvg/pyceres) built from source at tag v1.0.
+- Python >= 3.9, we recommend to use the virtual environment of your choice.
```
I ran into some troubles with Python > 3.10 + pycolmap 0.4.0. Maybe we can say here Python 3.9 / 3.10.

Similarly, miniconda created some headaches due to imcompatible system libs on Ubuntu 24 so maybe we shoul recommend python venv directly :)

ovysotska replied:
Ah, I see. I used venv.
Sure, I can change to more strict Python version.

ovysotska replied:
done.

ovysotska commented:
@pablospe thanks for the comment.

I checked the `docker/scripts` you recommend. They seems to provide instructions to installing `hloc` and dev dependencies.
In this PR, I am proposing to install the 'core dependencies' such that `pcd meshing` and `raybender` are not the part of the script. So hopefully, addressing your concerns here.

About the docker image, yes I tried buidling it. The building part seemed to work :) at least there were no obvious errors.
Unfortunately, I don't know what to do next with it. The readme doesn't tell how to use this image to run laMAR :cry: 
I tried creating  a container with this image, however, it just did nothing.
If you could add some instructions how to use that image, that would be great :smile: and I could try that out too.

Thanks!

pablospe commented:
> In this PR, I am proposing to install the 'core dependencies' such that pcd meshing and raybender are not the part of the script. 

Ok, I see. Anyways I would like to split this script in several scripts that could be reusable also in the Dockerfile. I will do/propose some changes to this PR, but essentially the same code. Thanks for the contribution!

ovysotska commented:
I will try but now I have a question. Do I need to create my own virtual environment, for example through `venv` before calling `.install_core_dependencies.sh` script?

pablospe commented:
> I will try but now I have a question. Do I need to create my own virtual environment, for example through `venv` before calling `.install_core_dependencies.sh` script?

Yes, if you want to install the python packages in a new environment. Otherwise, it will install them in the user's default one.

pablospe commented:
> `.install_core_dependencies.sh` script

Notice that now all the bash scripts are in `scripts/` folder. Then, `scripts/install_core_dependencies.sh` for running it.

ovysotska commented on scripts/install_core_dependencies.sh:
```diff
@@ -17,4 +17,4 @@ sudo ${root_folder}/scripts/install_colmap.sh
 
 # Pyceres.
-python3 -m pip install git+https://github.com/cvg/pyceres.git@v1.0
+python3 -m pip install git+https://github.com/cvg/pyceres.git@v2.3
```
@pablospe did you try running the full pipeline on the test data?

I was installing latest version of pyceres in my first iterations and it was building but when I ran the pipeline `lamar.run` on the CAB data it was crashing in the middle.

Just want to warn you that this happened to me before.

pablospe replied:
ok, let me revert this change.

pablospe replied:
@mihaidusmanu do you know what part isn't working? Wondering if it would be easy to update the code to make work with last version.

pablospe replied:
```suggestion
python3 -m pip install git+https://github.com/cvg/pyceres.git@v1.0
```

mihaidusmanu replied:
@pablospe not sure, but I wouldn't change any of the original versions without 1. rerunning pseudo-GT pipeline on some data 2. running some evals on the lamar release. There's quite a few things that can go wrong.

pablospe commented on Dockerfile:
```diff
@@ -123,7 +123,7 @@ RUN apt-get install -y --no-install-recommends --no-install-suggests wget && \
 
 # Build pyceres.
-RUN git clone --depth 1 --recursive https://github.com/cvg/pyceres
+RUN git clone --depth 1 -b v2.3 --recursive https://github.com/cvg/pyceres
```
This will fail, probably I need to also install Colmap from source explicitly here. With the latest version of pyceres this wasn't needed.
```suggestion
RUN git clone --depth 1 -b v1.0 --recursive https://github.com/cvg/pyceres
```

pablospe commented on scripts/build_hloc.sh:
```diff
@@ -0,0 +1,13 @@
+cd hloc
+
+# Build the wheel.
+pip wheel --no-deps -w dist-wheel .
+whl_path=$(find dist-wheel/ -name "*.whl")
+echo $whl_path >dist-wheel/whl_path.txt
```
I see. Thanks for point this out!

This separation into `builder` and `runtime` Docker image stages was done to exclude development dependencies from the runtime image, thus reducing the overall image size. It appears to be built this way, but I need to verify what you're saying. I didn't actually use the Lamar image but rather the Scantools one.

Notice that the script introduced in this PR (which installs it locally) doesn't use this method. Instead, it directly runs:
`python3 -m pip install git+https://github.com/cvg/Hierarchical-Localization.git@v1.4`
Therefore, this issue only affects the Docker image.

pablospe replied:
> We generally build with the `-e` option and can't turn this into a wheel.

Do you have a suggestion for a quick test I can run to verify that hloc is correctly installed? At least this creates the wheel successfully, probably without the feature extractors/matchers, as you said.

pablospe replied:
ok, thanks again. I will fix it soon.

pablospe commented:
I believe it should be ready for merging.

mihaidusmanu commented on README.md:
```diff
@@ -66,24 +66,29 @@ T_w_i = sessions.trajectories[keys[0]]  # first pose, from sensor/rig to world
+  * [Ceres Solver 2.1](https://ceres-solver.googlesource.com/ceres-solver/+/refs/tags/2.1.0)
+  * [Colmap 3.8](https://colmap.github.io/install.html) built from source. Note: **Do not install libceres-dev** as it was installed in the previous step.
+  * [hloc 1.4](https://github.com/cvg/Hierarchical-Localization) and its dependencies
+  * [pyceres v1.0](https://github.com/cvg/pyceres) built from source.
```
nit: not needed since mentioned in pyproject.

mihaidusmanu commented on scripts/install_core_dependencies.sh:
```diff
@@ -0,0 +1,22 @@
+${root_folder}/scripts/install_hloc.sh
+
+# Pyceres.
+python3 -m pip install git+https://github.com/cvg/pyceres.git@v1.0
```
Same, can be removed (installed in next step)

pablospe commented:
> thanks for your work. Is there something I could do about this PR?

Thanks, Olga. I will merge it when it passes the CI checks.
